apps/qubit/modules/informationobject/actions/calculateDatesAction.class.php

```diff
@@ -0,0 +1,151 @@
+        {
+          // Otherwise, display the events using a SELECT element
+          $this->form->setWidget($name, new sfWidgetFormSelect(array('choices' => $this->events)));
+          $label = $this->i18n->__('Event');
+        }
+
+        $this->form->setValidator($name, new sfValidatorInteger);
+        $this->form->getWidgetSchema()->$name->setLabel($label);
+
+        break;
+    }
+  }
+
+  protected function processField($field)
+  {
+    switch ($name = $field->getName())
+    {
+      case 'eventId':
+        $this->eventId = $field->getValue();
+
+        break;
+    }
+  }
+
+  protected function earlyExecute()
+  {
+    $this->i18n = $this->context->i18n;
+    $this->resource = $this->getRoute()->resource;
+    $this->events = self::getResourceEventsWithDateRangeSet($this->resource);
+  }
+
+  public function execute($request)
+  {
+    parent::execute($request);
+
+    // Redirect if unauthorized
+    if (!QubitAcl::check($this->resource, 'update'))
+    {
+      QubitAcl::forwardUnauthorized();
+    }
+
+    // Set response to 403 forbidden if attempting to calculate dates using
+    // non-existant descendants
+    if (!count($this->resource->descendants))
+    {
+      $this->getResponse()->setStatusCode(403);
+      return sfView::NONE;
+    }
+
+    if ($request->isMethod('post'))
+    {
+      $this->form->bind($request->getPostParameters());
+
+      if ($this->form->isValid())
+      {
+        $this->processForm();
+        $this->beginDateCalculation();
+        $this->redirect(array($this->resource, 'module' => 'informationobject'));
+      }
+    }
+
+    $message = $this->i18n->__('Warning: Selected date range for the specified event will be overwritten.');
+    $this->getUser()->setFlash('notice', $message);
+  }
+
+  protected function beginDateCalculation()
+  {
+    // Specify parameters for job
+    $params = array(
+      'objectId' => $this->resource->id,
+      'eventId' => $this->eventId
+    );
+
+    // Catch no Gearman worker available exception
+    // and others to show alert with exception message
+    try
+    {
+      QubitJob::runJob('arCalculateDescendantDatesJob', $params);
+
+      $message = $this->i18n->__('Date calculation started.');
+      $this->context->user->setFlash('info', $message);
+    }
+    catch (Exception $e)
+    {
+      $message = $this->i18n->__('Calculation failed') .': '. $this->i18n->__($e->getMessage());
+      $this->context->user->setFlash('error', $message);
+    }
+  }
+
+  static public function getResourceEventsWithDateRangeSet($resource)
+  {
+    $events = array();
+
+    $criteria = new Criteria;
+    $criteria->add(QubitEvent::OBJECT_ID, $resource->id);
+
+    // Assemble array of descriptions for any events containing date information
+    foreach(QubitEvent::get($criteria) as $event)
+    {
+      if ((!empty($event->date) || !empty($event->startDate) || !empty($event->endDate))
+        && (null !== $eventTypeTerm = QubitTerm::getById($event->typeId)))
```

Comment:
Sorry, I didn't express myself clear enough (or mixed two things in the same feedback) ...

You don't need to get the event by id, you already have it hydrated by the ORM as `type` (without Id) so, `null !== $event->typeId` in here and `$eventTypeName = $event->type->getName(array('cultureFallback' => true));` bellow should work.

Reply:
Ah, that makes sense. I'll fix that.